### PR TITLE
Subscriptions: add a floating subscribe button

### LINKED ISSUE #39772: Subscriptions: add a floating subscribe button
Add a floating subscribe button, which, when pressed, opens the subscribe modal:

We now can render "button only" subscribe block variation, and the email will be asked in a modal.

The floating button would be enabled via "Jetpack -> Settings -> Newsletter" alongside popups and other placements:

The button should be pulled from a template part, so that we can let folks jump in and modify the button's appearance, add some copy around it if they wish, etc.

Currently, this would clash with the "action bar" on WP.com sites:

We should keep showing the button even when subscribed when site has paid subscriptions; the button should then nudge to upgrade.

### PR DESCRIPTION
**Pull request by simison**
Resolves https://github.com/Automattic/jetpack/issues/39772

Adds floating subscribe button, which, when pressed, opens the subscribe modal. 

The button appearance can be modified directly in the site editor via the template part.

## Proposed changes:

Setting:

Website:

Screenshot 2024-06-05 at 21 17 32

image

### Other information:

In a follow-up PR:
- Keep showing when subscribed, but they could upgrade to paid subscription.

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:

* Either change `get_option( 'jetpack_subscribe_floating_button_enabled', false )` bit to `true` or add option to your site via CLI.
* Open frontend of the site.

### CODE REVIEW

File: projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		return self::$instance;
+	}
+
+	const BLOCK_TEMPLATE_PART_SLUG = 'jetpack-subscribe-floating-button';
+
+	/**
+	 * Jetpack_Subscribe_Floating_Button class constructor.
+	 */
+	public function __construct() {
+		if ( get_option( 'jetpack_subscribe_floating_button_enabled', true ) ) {
```
Comment (pkuliga) [inline projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php]: Let's disable it by default.

```suggestion
		if ( get_option( 'jetpack_subscribe_floating_button_enabled', false ) ) {
```

**Review by pkuliga — COMMENTED** [review]
Maybe worth adding some kind of an outline? Looks weird when the content beneath is black:

Screenshot 2024-06-10 at 16 42 31

**Comment by pkuliga** [comment]
When I close the subscribe modal the floating button is still highlighted:

Screenshot 2024-06-10 at 16 55 48

**Comment by simison** [comment]
> Maybe worth adding some kind of an outline? Looks weird when the content beneath is black:

TBH I'm not sure this would be too big issue; default button style would play well with default background, and inverse-colored sections (often footers) usually consist smaller portion of the website's surface. Adding custom styles would quickly get complicated with all the variations (light buttons against light backgrounds, dark buttons against dark backgrounds, and everything in-between) @crisbusquets thoughts?

**Comment by crisbusquets** [comment]
> > Maybe worth adding some kind of an outline? Looks weird when the content beneath is black:
> 
> TBH I'm not sure this would be too big issue; default button style would play well with default background, and inverse-colored sections (often footers) usually consist smaller portion of the website's surface. Adding custom styles would quickly get complicated with all the variations (light buttons against light backgrounds, dark buttons against dark backgrounds, and everything in-between) @crisbusquets thoughts?

I get why @pkuliga suggested it (it's almost invisible when the theme's background is black), but I agree with you, @simison. It would be difficult to find an outline that works for all the themes and it's not confused with the `focus` state.

File: projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+	/**
+	 * Makes get_block_template return the WP_Block_Template for the floating Subscribe button.
+	 *
+	 * @param WP_Block_Template $block_template The block template to be returned.
+	 * @param string            $id Template unique identifier (example: theme_slug//template_slug).
+	 * @param string            $template_type Template type: `'wp_template'` or '`wp_template_part'`.
+	 *
+	 * @return WP_Block_Template|null
+	 */
+	public function get_block_template_filter( $block_template, $id, $template_type ) {
```
Comment (lezama) [inline projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php:72]: As a follow up, we should probably have a parent class that implements these so we don't need to reimplement each time.

File: projects/plugins/jetpack/modules/subscriptions/subscribe-overlay/subscribe-overlay.css [hunk 3]
```diff
@@ -7,7 +7,7 @@ body.jetpack-subscribe-overlay-open {
 
 	visibility: hidden;
 	position: fixed;
-	z-index: 50000; /* Same as WP.com Action bar */
+	z-index: 50001; /* WP.com Action bar and floating subscribe button are 5000 */
```
Comment (lezama) [inline projects/plugins/jetpack/modules/subscriptions/subscribe-overlay/subscribe-overlay.css:10]:
```suggestion
	z-index: 50001; /* WP.com Action bar and floating subscribe button are 50000 */
```

File: projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php (hunk 2 above)
Comment (simison) [reply projects/plugins/jetpack/modules/subscriptions/subscribe-floating-button/class-jetpack-subscribe-floating-button.php:72]: Yasss: 
- https://github.com/Automattic/jetpack/issues/39853

File: projects/plugins/jetpack/modules/subscriptions/subscribe-overlay/subscribe-overlay.css (hunk 3 above)
Comment (simison) [reply projects/plugins/jetpack/modules/subscriptions/subscribe-overlay/subscribe-overlay.css:10]: This won't matter anymore now that we disable the Action bar when the floating button is enabled. :-)

**Comment by lezama** [comment]
Looking good! 🚢 IT

Screenshot 2024-10-21 at 2 43 50 PM

**Review by lezama — APPROVED** [review]

**Comment by simison** [comment]
One small todo for a follow up:

> We should keep showing the button even when subscribed when site has paid subscriptions; the button should then nudge to upgrade.

**Comment by crisbusquets** [comment]
Do we have tracking here? It would be great to have a Dashboard with the number of clicks.
cc @lezama

**Comment by lezama** [comment]
yes we do! https://github.com/Automattic/jetpack/pull/39860

**Comment by crisbusquets** [comment]
> yes we do! #39860

Thanks, @lezama! Then I should be able to create a Tracks funnel, right? 🤔

**Comment by lezama** [comment]
@crisbusquets yes, you will need to look for events like `wpcom_memberships_checkout_view` with the property `app_source === subscribe-floating-button`

**Pull request merged**